File: app/assets/javascripts/Patient.js
const mongoose = require('mongoose');
const Code = require('./basetypes/Code');
const Interval = require('./basetypes/Interval');
const Quantity = require('./basetypes/Quantity');
const DateTime = require('./basetypes/DateTime');
const AllDataElements = require('./AllDataElements');

const [Schema, Number, String] = [
  mongoose.Schema,
  mongoose.Schema.Types.Number,
  mongoose.Schema.Types.String,
];

const PatientSchema = new Schema({
  birthDatetime: DateTime,
  qdmVersion: { type: String, default: '5.3' },
  _type: { type: String, default: 'Patient' },

  givenNames: [String],
  familyName: String,
  bundleId: String,
  // These are the "data criteria", or QDM datatype elements that exist on a
  // patient.
  dataElements: [],
  // This field is for application specific information only. If both Bonnie
  // Cypress use a common field, it should be made a field on this model,
  // and not put into extendedData.
  extendedData: {
    type: Mixed,
    default: {},
  },
}, { id: false });

// After initialization of a Patient model, initialize every individual data element
// to its respective Mongoose Model
PatientSchema.methods.initializeDataElements = function initializeDataElements() {
  let typeStripped;
  const dataElementsInit = [];
  this.dataElements.forEach((element) => {
    typeStripped = element._type.replace(/QDM::/, '');
    if (typeStripped in AllDataElements) {
      dataElementsInit.push(AllDataElements[typeStripped](element));
    } else {
      dataElementsInit.push(element);
    }
  });
  this.set({ dataElements: dataElementsInit });
};

PatientSchema.queue('initializeDataElements');

PatientSchema.methods.id = function id() {
  return this._id;
};

// Returns an array of elements that exist on this patient, that
// match the given HQMF data criteria OID.
PatientSchema.methods.getByHqmfOid = function getByHqmfOid(hqmfOid) {
  return this.dataElements.filter(element => element.hqmfOid === hqmfOid);
};

// Returns an array of elements that exist on this patient, that
// match the given QRDA data criteria OID.
PatientSchema.methods.getByQrdaOid = function getByQrdaOid(qrdaOid) {
  return this.dataElements.filter(element => element.qrdaOid === qrdaOid);
};

// Returns an array of elements that exist on this patient. Optionally
// takes a category, which returns all elements of that QDM category.
// Example: patient.getDataElements(category = 'encounters') will return
// all Encounter QDM data types active on the patient.
PatientSchema.methods.getDataElements = function getDataElements(params) {
  if (params.category && params.qdmStatus) {
    return this.dataElements.filter(element => (element.category === params.category) && (element.qdmStatus === params.qdmStatus));
  } else if (params.category) {
    return this.dataElements.filter(element => element.category === params.category);
  }
  return this.dataElements;
};

// Returns an array of dataElements that exist on the patient, queried by
// QDM profile
PatientSchema.methods.getByProfile = function getByProfile(profile, isNegated = null) {
  // If isNegated == true, only return data elements with a negationRationale that isn't null.
  // If isNegated == false, only return data elements with a null negationRationale.
  // If isNegated == null, return all matching data elements by type, regardless of negationRationale.
  const results = this.dataElements.filter(element => element._type === `QDM::${profile}` && (isNegated === null || !!element.negationRationale === isNegated));
  return results.map(result => AllDataElements[profile](result));
};

// This method is called by the CQL execution engine on a CQLPatient when
// the execution engine wants information on a record. A record could be patient
// characteristic information about the patient, or it could be data criteria
// that currently exist on this patient (data criteria you drag on a patient
// in Bonnie's patient builder).
// @param {String} profile - the data criteria requested by the execution engine
// @returns {Object}
PatientSchema.methods.findRecords = function findRecords(profile) {
  let profileStripped;
  if (profile === 'Patient') {
    // Requested generic patient info
    const info = { birthDatetime: this.birthDatetime };
    return [info];
  } else if (/PatientCharacteristic/.test(profile)) {
    // Requested a patient characteristic
    profileStripped = profile.replace(/ *\{[^)]*\} */g, '');
    return this.getByProfile(profileStripped);
  } else if (profile != null) {
    // Requested something else (probably a QDM data type).

    // Strip model details from request. The requested profile string contains
    // a lot of things we don't need or care about. Example, we might see
    // something like:
    // "{urn:healthit-gov:qdm:v5_0_draft}PatientCharacteristicEthnicity"
    // Where we only care about: "PatientCharacteristicEthnicity".
    profileStripped = profile.replace(/ *\{[^)]*\} */g, '');

    // Check and handle negation status
    if (/Positive/.test(profileStripped)) {
      profileStripped = profileStripped.replace(/Positive/, '');
      // Since the data criteria is 'Positive', it is not negated.
      return this.getByProfile(profileStripped, false);
    } else if (/Negative/.test(profileStripped)) {
      profileStripped = profileStripped.replace(/Negative/, '');
      // Since the data criteria is 'Negative', it is negated.
      return this.getByProfile(profileStripped, true);
    }
    // No negation status, proceed normally
    return this.getByProfile(profileStripped);
  }
  return [];
};

PatientSchema.methods.adverse_events = function adverse_events() {
  return this.getDataElements({ category: 'adverse_event' });
};

PatientSchema.methods.allergies = function allergies() {
  return this.getDataElements({ category: 'allergy' });
};

PatientSchema.methods.assessments = function assessments() {
  return this.getDataElements({ category: 'assessment' });
};

PatientSchema.methods.care_experiences = function care_experiences() {
  return this.getDataElements({ category: 'care_experience' });
};

PatientSchema.methods.care_goals = function care_goals() {
  return this.getDataElements({ category: 'care_goal' });
};

PatientSchema.methods.communications = function communications() {
  return this.getDataElements({ category: 'communication' });
};

PatientSchema.methods.conditions = function conditions() {
  return this.getDataElements({ category: 'condition' });
};

PatientSchema.methods.devices = function devices() {
  return this.getDataElements({ category: 'device' });
};

PatientSchema.methods.diagnostic_studies = function diagnostic_studies() {
  return this.getDataElements({ category: 'diagnostic_study' });
};

PatientSchema.methods.encounters = function encounters() {
  return this.getDataElements({ category: 'encounter' });
};

PatientSchema.methods.family_history = function family_history() {
  return this.getDataElements({ category: 'family_history' });
};

PatientSchema.methods.functional_statuses = function functional_statuses() {
  return this.getDataElements({ category: 'functional_status' });
};

PatientSchema.methods.immunizations = function immunizations() {
  return this.getDataElements({ category: 'immunization' });
};

PatientSchema.methods.interventions = function interventions() {
  return this.getDataElements({ category: 'intervention' });
};

PatientSchema.methods.laboratory_tests = function laboratory_tests() {
  return this.getDataElements({ category: 'laboratory_test' });
};

PatientSchema.methods.medical_equipment = function medical_equipment() {
  return this.getDataElements({ category: 'medical_equipment' });
};

PatientSchema.methods.medications = function medications() {
  return this.getDataElements({ category: 'medication' });
};

PatientSchema.methods.physical_exams = function physical_exams() {
  return this.getDataElements({ category: 'physical_exam' });
};

PatientSchema.methods.preferences = function preferences() {
  return this.getDataElements({ category: 'preference' });
};

PatientSchema.methods.provider_characteristics = function provider_characteristics() {
  return this.getDataElements({ category: 'provider_characteristic' });
};

PatientSchema.methods.procedures = function procedures() {
  return this.getDataElements({ category: 'procedure' });
};

PatientSchema.methods.results = function results() {
  return this.getDataElements({ category: 'result' });
};

PatientSchema.methods.risk_category_assessments = function risk_category_assessments() {
  return this.getDataElements({ category: 'risk_category_assessment' });
};

PatientSchema.methods.social_history = function social_history() {
  return this.getDataElements({ category: 'social_history' });
};

PatientSchema.methods.substances = function substances() {
  return this.getDataElements({ category: 'substance' });
};

PatientSchema.methods.symptoms = function symptoms() {
  return this.getDataElements({ category: 'symptom' });
};

PatientSchema.methods.system_characteristics = function system_characteristics() {
  return this.getDataElements({ category: 'system_characteristic' });
};

PatientSchema.methods.transfers = function transfers() {
  return this.getDataElements({ category: 'transfer' });
};

PatientSchema.methods.vital_signs = function vital_signs() {
  return this.getDataElements({ category: 'vital_sign' });
};

module.exports.PatientSchema = PatientSchema;
module.exports.Patient = mongoose.model('Patient', PatientSchema);

File: app/models/qdm/patient.rb
module QDM
  # app/models/qdm/patient.rb
  class Patient
    include Mongoid::Document
    field :birthDatetime, type: DateTime
    field :qdmVersion, type: String, default: '5.5'
    embedded_in :cqmPatient, class_name: 'CQM::Patient'

    # These are the "data criteria", or QDM datatype elements that exist on a
    # patient.
    embeds_many :dataElements, class_name: 'QDM::DataElement'

    # This field is for application specific information only. If both Bonnie
    # Cypress use a common field, it should be made a field on this model,
    # and not put into extendedData.
    field :extendedData, type: Hash

    # Returns an array of elements that exist on this patient, that
    # match the given HQMF data criteria OID.
    def get_by_hqmf_oid(hqmf_oid)
      dataElements.where(hqmfOid: hqmf_oid) || []
    end

    # Returns an array of elements that exist on this patient, that
    # match the given QRDA data criteria OID.
    def get_by_qrda_oid(qrda_oid)
      dataElements.where(qrdaOid: qrda_oid) || []
    end

    # Shift all data element fields that deal with dates by the given value.
    # Given value should be in seconds. Positive values shift forward, negative
    # values shift backwards.
    #
    # Note: This will shift dates of the birthdate and
    # dates on the data elements that exist on the patient.
    def shift_dates(seconds)
      self.birthDatetime = (birthDatetime.utc.to_time + seconds.seconds).to_datetime
      dataElements.each { |element| element.shift_dates(seconds) }
    end

    # Returns an array of elements that exist on this patient. Optionally
    # takes a category and/or, which returns all elements of that QDM
    # category. Example: patient.get_data_elements('encounters')
    # will return all Encounter QDM data types active on the patient.
    def get_data_elements(category = nil, status = nil)
      if category && status
        dataElements.where(qdmCategory: category, qdmStatus: status) || []
      elsif category
        dataElements.where(qdmCategory: category) || []
      else
        dataElements || []
      end
    end

    # Helper method; returns adverse_event data element types on this patient.
    def adverse_events
      get_data_elements('adverse_event')
    end

    # Helper method; returns allergy data element types on this patient.
    def allergies
      get_data_elements('allergy')
    end

    # Helper method; returns assessment data element types on this patient.
    def assessments
      get_data_elements('assessment')
    end

    # Helper method; returns care_experience data element types on this patient.
    def care_experiences
      get_data_elements('care_experience')
    end

    # Helper method; returns care_goal data element types on this patient.
    def care_goals
      get_data_elements('care_goal')
    end

    # Helper method; returns communication data element types on this patient.
    def communications
      get_data_elements('communication')
    end

    # Helper method; returns condition data element types on this patient.
    def conditions
      get_data_elements('condition')
    end

    # Helper method; returns device data element types on this patient.
    def devices
      get_data_elements('device')
    end

    # Helper method; returns diagnostic_study data element types on this patient.
    def diagnostic_studies
      get_data_elements('diagnostic_study')
    end

    # Helper method; returns encounter data element types on this patient.
    def encounters
      get_data_elements('encounter')
    end

    # Helper method; returns family_history data element types on this patient.
    def family_history
      get_data_elements('family_history')
    end

    # Helper method; returns functional_status data element types on this patient.
    def functional_statuses
      get_data_elements('functional_status')
    end

    # Helper method; returns immunization data element types on this patient.
    def immunizations
      get_data_elements('immunization')
    end

    # Helper method; returns intervention data element types on this patient.
    def interventions
      get_data_elements('intervention')
    end

    # Helper method; returns laboratory_test data element types on this patient.
    def laboratory_tests
      get_data_elements('laboratory_test')
    end

    # Helper method; returns medical_equipment data element types on this patient.
    def medical_equipment
      get_data_elements('medical_equipment')
    end

    # Helper method; returns medication data element types on this patient.
    def medications
      get_data_elements('medication')
    end

    # Helper method; returns patient_characteristic data element types on this patient.
    def patient_characteristics
      get_data_elements('patient_characteristic')
    end

    # Helper method; returns physical_exam data element types on this patient.
    def physical_exams
      get_data_elements('physical_exam')
    end

    # Helper method; returns preference data element types on this patient.
    def preferences
      get_data_elements('preference')
    end

    # Helper method; returns provider_characteristic data element types on this patient.
    def provider_characteristics
      get_data_elements('provider_characteristic')
    end

    # Helper method; returns procedure data element types on this patient.
    def procedures
      get_data_elements('procedure')
    end

    # Helper method; returns result data element types on this patient.
    def results
      get_data_elements('result')
    end

    # Helper method; returns risk_category_assessment data element types on this patient.
    def risk_category_assessments
      get_data_elements('risk_category_assessment')
    end

    # Helper method; returns social_history data element types on this patient.
    def social_history
      get_data_elements('social_history')
    end

    # Helper method; returns substance data element types on this patient.
    def substances
      get_data_elements('substance')
    end

    # Helper method; returns symptom data element types on this patient.
    def symptoms
      get_data_elements('symptom')
    end

    # Helper method; returns system_characteristic data element types on this patient.
    def system_characteristics
      get_data_elements('system_characteristic')
    end

    # Helper method; returns transfer data element types on this patient.
    def transfers
      get_data_elements('transfer')
    end

    # Helper method; returns vital_sign data element types on this patient.
    def vital_signs
      get_data_elements('vital_sign')
    end

    # Return the Mongo id.
    def id
      _id
    end

    # Include '_type' in any JSON output. This is necessary for deserialization.
    def to_json(options = nil)
      serializable_hash(methods: :_type, include: :_type).to_json(options)
    end
  end
end
